### DISCUSSION TITLE
Is it possible to run Nextflow with Docker and a Conda environment inside a Docker?

### DISCUSSION THREAD

**Discussion by hkarakurt8742** [nextflow-io/nextflow#3539] [General]
Hello everyone,
I am using Nextflow for my local pipelines and I want to move to Docker but I have a problem. My pipeline includes a software that needs to be installed as DEB package. Other packages have conda versions but I need to install them to another environment. Is it possible to define a Docker image and a Conda environment inside Docker to run a Nextflow script?

Thank you in advance

**Comment by mribeirodantas** [comment]
Hello, @hkarakurt8742 ! It's slightly confusing what you wrote, so depending on what you mean, yes or no 😅 

Yes, you can have a Docker image whose set of software within it is installed/managed with conda. Actually, that's pretty common! You can even do that with [Wave](https://seqera.io/wave/) (check [here](https://github.com/seqeralabs/wave-showcase/tree/master/example2)).

However, if you want to have conda+docker inside a docker container (a Docker container running inside a Docker container), I'd strongly recommend you to not do so. It's not that it's impossible (check `-dockerize` in Nextflow), it's just that it's prone to a lot of headaches and issues.

Maybe you have a corner case, and such a thing is required, but usually, it's not. Please, share more details if you think otherwise.

**Reply by hkarakurt8742** [reply]
Hello @mribeirodantas! It is good to see you here 😄. 
Maybe I should clarify my question. I created a docker image (Ubuntu 20.04) and installed conda in it. Then I created 2 different conda environments (Env1 and Env2). Env1 has Tensorflow and it conflicts with some of the packages in Env2. I can run a Nextflow script with Docker, it is going well but is it possible (or feasible) to activate/deactivate Conda environments that built in a Docker image?

 Thank you for your answer.

**Reply by mribeirodantas** [reply]
Hello, Hamza! Amazing to see you here 😄 

Before answering your question, it's important to mention that it's really good practice to run Nextflow in a real machine and have a single container image for every process. If you have one container image for all processes, it can not only be trickier to debug in case you run into issues, but the image will be larger to `docker pull` and slower to `docker run` (and remember this is done for every task, so 2 processes with 10 samples may mean 20 docker pulls/runs). Depending on your executor, and how your pipeline is assembled, maybe you won't have to pull every time, but it could be the case on the cloud, for example. So one container image, with conda managing the installation of files if you prefer (I would recommend micromamba here, for it is faster) would apparently be a solution for your situation (Check section `4.3.2` [here](https://training.seqera.io/#_condabioconda_packages)). If you really want to run Nextflow with Conda and without Docker, but inside a Docker container, you can do that. Simply set the `conda` directive to point to the environment you want (Env1 or Env2), but I insist this is not a good practice 😅 There are many reported issues when running Docker within Docker (regardless of Nextflow).

**Reply by hkarakurt8742** [reply]
Thank you for your detailed answer. I will try all combinations and find the best solutions but I think keeping it simple inside the Docker will be a better option :)
